### PR TITLE
compiler: Revamp CIRE exploiting EvalDerivative

### PR DESCRIPTION
**Pull request by FabioLuporini**
By searching and optimizing EvalDerivatives, that is sub-expressions produced by the evaluation of Derivatives, the CIRE pass is now more rigorous, robust, effective, and quick. It also opens up the possibility of easily synthesizing different variants of the same code.

The PR is equipped with many new tests, and adjust all of the existing ones to the new CIRE infrastructure.

Some optimization options , such as `cire-mincost-*` and `cire-repeats-*` have finally been dropped and replaced by two simple knobs, namely `cire-mingain` and `cire-schedule`, which are suitably documented

EDIT: I will also leave some inline comments for ease of reviewing 

Note 1: AFAICT, there is no compilation speed regression

Note 2: this PR may impact the code generated by the compiler for some Operators. In all my experiments, I have not seen even a single regression in execution time.

So, Note 1 + Note 2 => all sort of timings should be improving. It'd be great to have TheMatrix to easily test this on multiple architectures simultaneously, but for now we'll need some beta testing I guess

### CODE REVIEW

File: devito/core/cpu.py [hunk 1]
```diff
@@ -173,18 +165,16 @@ def _specialize_clusters(cls, clusters, **kwargs):
...
         clusters = extract_increments(clusters, sregistry)
         clusters = cire(clusters, 'sops', sregistry, options, platform)
         clusters = factorize(clusters)
         clusters = optimize_pows(clusters)
 
-        # The previous passes may have created fusion opportunities, which in
-        # turn may enable further optimizations
+        # The previous passes may have created fusion opportunities
         clusters = fuse(clusters)
-        clusters = eliminate_arrays(clusters)
```
Comment (FabioLuporini) [inline devito/core/cpu.py:185]: Note for reviewers: yep, we can drop a whole pass, as thanks to the many generalisations it has become unnecessary (while being much weaker then what we have now)

File: devito/core/gpu.py [hunk 2]
```diff
@@ -156,19 +149,17 @@ def _specialize_clusters(cls, clusters, **kwargs):
...
-        clusters = cse(clusters, sregistry)
-
-        # Lifting may create fusion opportunities, which in turn may enable
-        # further optimizations
+        # The previous passes may have created fusion opportunities
         clusters = fuse(clusters)
-        clusters = eliminate_arrays(clusters)
+
+        # Reduce flops
+        clusters = cse(clusters, sregistry)
```
Comment (FabioLuporini) [inline devito/core/gpu.py:162]: Note for reviewers: note that now `cse` is performed after `fuse`. That's fine, it was wrong before

File: devito/finite_differences/derivative.py [hunk 3]
```diff
@@ -359,8 +356,4 @@ def _eval_fd(self, expr):
         for e in self._ppsubs:
             res = res.xreplace(e)
 
-        # Step 5: Cast to EvaluatedDerivative
-        assert res.is_Add
-        res = EvalDerivative(*res.args, evaluate=False)
```
Comment (FabioLuporini) [inline devito/finite_differences/derivative.py:364]: Note for reviewers: now unnecessary as done directly inside finite_difference.py

File: devito/ir/clusters/queue.py [hunk 4]
```diff
@@ -117,24 +117,3 @@ def _fetch_properties(self, clusters, prefix):
             for i in prefix:
                 properties[i.dim].update(v.get(i.dim, set()))
         return properties
-
-
-class Context(Queue):
```
Comment (FabioLuporini) [inline devito/ir/clusters/queue.py:122]: Note for reviewers: dropping as unnecessary now

File: devito/passes/clusters/factorization.py [hunk 5]
```diff
@@ -78,8 +78,11 @@ def collect_const(expr):
             # Back to the running example
             # -> (a + c)
             add = Add(*v)
-            # -> 3.*(a + c)
-            mul = Mul(k, add, evaluate=False)
+            if add == 0:
```
Comment (FabioLuporini) [inline devito/passes/clusters/factorization.py:81]: Note for reviewers: these are small bug fixes... without, in some corner circumstances we were generating expressions such as `a + 0`, while now we would generate just `a`

File: devito/passes/equations/linearity.py [hunk 6]
```diff
@@ -185,7 +185,7 @@ def _(expr):
 
     derivs, others = split(args, lambda a: isinstance(a, sympy.Derivative))
     if not derivs:
-        return expr
+        return reuse_if_untouched(expr, args)
```
Comment (FabioLuporini) [inline devito/passes/equations/linearity.py:188]: Note for reviewers: this is a performance-bug fix, and it's now tested

File: devito/passes/clusters/aliases.py [hunk 7]
```diff
@@ -485,91 +543,89 @@ def collect(extracted, ispace, min_storage):
...
-        for i in cluster.ispace.intervals:
+        indicess = [[] for _ in a.distances]
+        for i in meta.ispace:
             try:
                 interval = imapper[i.dim]
             except KeyError:
-                # E.g., `x0_blk0` or (`a[y_m+1]` => `y not in imapper`)
-                intervals.append(i)
-                continue
+                if i.dim in a.free_symbols:
```
Comment (FabioLuporini) [inline devito/passes/clusters/aliases.py:616]: Note for reviewers: this is a small bug fix

File: devito/passes/clusters/aliases.py [hunk 8]
```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
...
 
-    def _selector(self, e, naliases):
-        if naliases <= 1:
-            return 0
-        else:
-            return estimate_cost(e, True)*naliases // self._opt_mincost
+    def _generate(self, exprs, exclude):
+        # E.g., extract `u.dx*a*b` and `u.dx*a*c` from `[(u.dx*a*b).dy`, `(u.dx*a*c).dy]`
+        def cbk_search(expr):
+            if isinstance(expr, EvalDerivative) and not expr.base.is_Function:
```
Comment (FabioLuporini) [inline devito/passes/clusters/aliases.py:331]: Note for reviewers: this is the game changer -- we can finally look for `EvalDerivative`, rather than speculatively searching all sort of expressions in sum-of-product form

**Review by mloubout — COMMENTED** [review]
Some comments. looks good

File: devito/passes/clusters/aliases.py [hunk 9]
```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
...
+        """
+        Create a key for the given Cluster. Clusters with same key may be
+        processed together in the search for CIRE candidates. Clusters should
+        have a different key if they must not be processed together, e.g.,
+        when this would lead to violation of data dependencies.
+        """
         raise NotImplementedError
 
-    def _in_writeto(self, dim, cluster):
+    def _betterv(self, delta_flops, delta_ws):
```
Comment (georgebisbas) [inline devito/passes/clusters/aliases.py:209]: name change?

Comment (georgebisbas) [reply devito/passes/clusters/aliases.py:209]: deliver typo

File: devito/passes/clusters/aliases.py [hunk 10]
```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
...
-            for e in exprs:
-                for i in search(e, self._rule, 'all', 'bfs_first_hit'):
-                    if {a.function for a in i.free_symbols} & exclude:
-                        continue
-                    mapper.add(i, self._make_symbol)
+        # Rule out extractions that would break data dependencies
+        exclude = set().union(*[c.scope.writes for c in clusters])
 
-        return mapper
+        # Rule out extractions that are *not* independent of the Dimension
```
Comment (georgebisbas) [inline devito/passes/clusters/aliases.py:242]: are not independent \equivalent to are dependent (?)

File: devito/passes/clusters/aliases.py [hunk 11]
```diff
@@ -485,91 +543,91 @@ def collect(extracted, ispace, min_storage):
                 distance = [(d, set(v)) for d, v in LabeledVector.transpose(*distance)]
                 distances.append(LabeledVector([(d, v.pop()) for d, v in distance]))
 
-            aliases.add(alias, list(mapper), aliaseds, distances)
+            # Compute the alias score
+            na = len(aliaseds)
```
Comment (georgebisbas) [inline devito/passes/clusters/aliases.py:547]: other var name?

File: devito/passes/clusters/aliases.py [hunk 12]
```diff
@@ -485,91 +543,91 @@ def collect(extracted, ispace, min_storage):
...
-        if score > 0:
-            candidates[e] = score
-            aliaseds.extend(v.aliaseds)
-        else:
-            others.append(e)
+    # `score < m` => discarded
+    # `score > M` => optimized
+    # `m <= score <= M` => maybe discarded, maybe optimized -- depends on heuristics
+    m = mingain
+    M = mingain*3
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply devito/passes/clusters/aliases.py:568]: heuristic?

File: devito/passes/clusters/aliases.py (hunk 7 above)
Comment (georgebisbas) [reply devito/passes/clusters/aliases.py:616]: What does it fix? What is the bug?

File: devito/symbolics/manipulation.py [hunk 13]
```diff
@@ -294,9 +294,10 @@ def subs_op_args(expr, args):
     return expr.subs({i.name: args[i.name] for i in expr.free_symbols if i.name in args})
 
 
-def rebuild_if_untouched(expr, args, evaluate=False):
+def reuse_if_untouched(expr, args, evaluate=False):
```
Comment (georgebisbas) [inline devito/symbolics/manipulation.py:297]: right

**Review by georgebisbas — COMMENTED** [review]
Will shortly do another pass

File: devito/core/cpu.py [hunk 14]
```diff
@@ -24,17 +22,17 @@ class Cpu64OperatorMixin(object):
...
-    optimized away. Dimension-invariant aliases are lifted outside of one or more
-    invariant loop(s), so they require tensor temporaries that can be potentially
-    very large (e.g., the whole domain in the case of time-invariant aliases).
+    Minimum operation count reduction for a redundant expression to be optimized
+    away. Higher (lower) values make a redundant expression less (more) likely to
+    be optimized away.
     """
 
-    CIRE_MINCOST_SOPS = 10
+    CIRE_SCHEDULE = 'automatic'
```
[the remark this answers is not in the document]
Comment (georgebisbas) [reply devito/core/cpu.py:32]: > Higher (lower) values
this higher lower seems a bit unclear to me

File: devito/ir/support/properties.py [hunk 15]
```diff
@@ -81,3 +81,7 @@ def normalize_properties(*args):
         properties.update(p - drop)
 
     return properties
+
+
+def relax_properties(properties):
```
Comment (georgebisbas) [inline devito/ir/support/properties.py:86]: I think you mentioned that in the past? To drop properties?

File: tests/test_dse.py [hunk 16]
```diff
@@ -1598,7 +1654,200 @@ def test_extraction_from_lifted_ispace(self, rotate):
...
+
+        op = Operator(eqn, subs=grid.spacing_map, openmp=True)
+
+        # Check code generation
+        assert op._profiler._sections['section1'].sops == exp_ops
+        arrays = [i for i in FindSymbols().visit(op._func_table['bf0']) if i.is_Array]
+        assert len(arrays) == 5
+        assert len(FindNodes(VExpanded).visit(op._func_table['bf0'])) == 3
+        xs, ys, zs = self.get_params(op, 'x0_blk0_size', 'y0_blk0_size', 'z_size')
+        # The three kind of derivatives taken -- in x, y, z -- all are over
```
Comment (georgebisbas) [inline tests/test_dse.py:1696]: three kinds?

File: tests/test_dse.py [hunk 17]
```diff
@@ -1708,12 +1959,13 @@ def test_tti_adjoint_akin_v2(self):
...
-        summary = op1(time_M=1, p=p1)
-        assert np.isclose(norm(p), norm(p1), atol=1e-15)
+        summary1 = op1(time_M=1, p=p1)
+        exp_p = norm(p)
+        assert np.isclose(exp_p, norm(p1), atol=1e-15)
 
         # Also check against expected operation count to make sure
         # all redundancies have been detected correctly
-        assert summary[('section1', None)].ops == 80
+        assert summary1[('section1', None)].ops == 75
```
Comment (georgebisbas) [inline tests/test_dse.py:1968]: WHy summary1 instead of summary?

File: tests/test_dse.py [hunk 18]
```diff
@@ -2236,19 +2542,30 @@ def test_fullopt_w_mpi(self):
...
+        assert op._profiler._sections['section1'].sops == expected
+        assert len([i for i in FindSymbols().visit(op) if i.is_Array]) == 7
+
 
 class TestTTIv2(object):
 
     @switchconfig(profiling='advanced')
     @pytest.mark.parametrize('space_order,expected', [
-        (4, 212), (12, 420)
+        (4, 191), (12, 383)
```
Comment (georgebisbas) [inline tests/test_dse.py:2568]: nice

File: devito/core/gpu.py [hunk 19]
```diff
@@ -69,15 +68,9 @@ def _normalize_kwargs(cls, **kwargs):
...
         o['cire-ftemps'] = oo.pop('cire-ftemps', False)
-        o['cire-mincost'] = {
-            'invariants': {
-                'scalar': 1,
-                'tensor': oo.pop('cire-mincost-inv', cls.CIRE_MINCOST_INV),
-            },
-            'sops': oo.pop('cire-mincost-sops', cls.CIRE_MINCOST_SOPS)
-        }
+        o['cire-mingain'] = oo.pop('cire-mingain', cls.CIRE_MINGAIN)
+        o['cire-schedule'] = oo.pop('cire-schedule', cls.CIRE_SCHEDULE)
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/core/gpu.py:73]: I thought about it (obviously) but in reality there are quite a few differences between the two, so I'd rather keep it separated

File: devito/finite_differences/differentiable.py [hunk 20]
```diff
@@ -363,12 +370,44 @@ def _eval_at(self, func):
...
 class Mul(DifferentiableOp, sympy.Mul):
     __sympy_class__ = sympy.Mul
-    __new__ = DifferentiableOp.__new__
+
+    def __new__(cls, *args, **kwargs):
+        # A DifferentiableOp may not trigger evaluation upon construction
+        # (e.g., if an EvalDerivative is present among the arguments)
+        # So we treat some special cases here
+
+        # a*0 -> 0
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:400]: this is never an issue to the best of my knowledge

I think the only issue was with EvalDerivative (and therefore Add) because they explicitly forbid evaluation

File: devito/finite_differences/differentiable.py [hunk 21]
```diff
@@ -411,17 +450,46 @@ def _gather_for_diff(self):
...
+        # a+0 -> a
+        args = [i for i in args if i != 0]
+
+        # Reorder for homogeneity with pure SymPy types
+        _addsort(args)
+
+        obj = super().__new__(cls, *args, **kwargs)
+
+        try:
+            obj.base = base
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:475]: CIRE won't optimize EvalDerivatives if the base is a pure Function, such as `u.dx`, simply because there would be nothing to optimize. It needs to be an expression to be worth optimizing, ie `expr.dx` . CTRL+F for `.base` and you'll see where it's used in `aliases.py`

File: devito/finite_differences/differentiable.py [hunk 22]
```diff
@@ -411,17 +450,46 @@ def _gather_for_diff(self):
...
+        except AttributeError:
+            # This might happen if e.g. one attempts a (re)construction with
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:477]: I tried, but I wasn't able to produce them, not sure why... any chance you can try writing a reproducer?

File: devito/passes/clusters/aliases.py (hunk 9 above)
Comment (FabioLuporini) [reply devito/passes/clusters/aliases.py:209]: agree, gonna find a better name, current one is leftover

File: devito/passes/clusters/aliases.py (hunk 10 above)
Comment (FabioLuporini) [reply devito/passes/clusters/aliases.py:242]: indeed, will fix

File: devito/passes/clusters/aliases.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/passes/clusters/aliases.py:568]: yes. Heuristics are in master too in this same function, though the "3" is new. It works perfectly for what we want and don't want to optimize given the other chosen values, what can I say..

File: devito/passes/clusters/aliases.py (hunk 7 above)
Comment (FabioLuporini) [reply devito/passes/clusters/aliases.py:616]: I don't remember exactly anymore, but I've added a test. It had to do with hierarchical blocking and maybe nested parallelism

File: devito/core/cpu.py [hunk 23]
```diff
@@ -24,17 +22,17 @@ class Cpu64OperatorMixin(object):
     3 => "blocks", "sub-blocks", and "sub-sub-blocks", ...
     """
 
-    CIRE_MINCOST_INV = 50
+    CIRE_MINGAIN = 10
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/core/cpu.py:25]: good point, I'm going to deprecate stuff and raise warnings accordingly

File: devito/ir/clusters/cluster.py [hunk 24]
```diff
@@ -164,7 +164,7 @@ def used_dimensions(self):
         example, reduction or redundant (i.e., invariant) Dimensions won't
         appear in an expression.
         """
-        return set().union(*[i._defines for i in self.free_symbols if i.is_Dimension])
+        return {i for i in self.free_symbols if i.is_Dimension}
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/ir/clusters/cluster.py:167]: oh, now it adheres to the specification of the method (docstring)
before, it was mistakenly adding _all_ dimensions, including the parent ones, which may actually not appear directly in the cluster

File: devito/finite_differences/differentiable.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:404]: @mloubout turn out it was an issue in a very corner case. I added flattening + test (a test that took me a while to write out...)

[the remark this answers is not in the document]
Comment (mloubout) [reply devito/finite_differences/differentiable.py:405]: 👍

File: devito/passes/clusters/aliases.py [hunk 25]
```diff
@@ -1150,14 +1234,50 @@ def cit(ispace0, ispace1):
...
+    Split potential derivative coefficients and arguments into two groups.
     """
-    if expr.is_Number:
-        return True
-    indexeds = [i for i in expr.free_symbols if i.is_Indexed]
-    return any(not set(grid.dimensions) <= set(i.function.dimensions) for i in indexeds)
+    # TODO: Once we'll be able to keep Derivative intact down to this point,
+    # we won't probably need this function anymore, because, in essence, what
+    # this function is doing is reconstructing prematurely lowered information
+    grids = {getattr(i.function, 'grid', None) for i in expr.free_symbols} - {None}
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/passes/clusters/aliases.py:1256]: ping

File: devito/operator/operator.py [hunk 26]
```diff
@@ -918,6 +918,14 @@ def parse_kwargs(**kwargs):
     options.setdefault('mpi', configuration['mpi'])
     for k, v in configuration['opt-options'].items():
         options.setdefault(k, v)
+    # Handle deprecations
+    deprecated_options = ('cire-mincost-inv', 'cire-mincost-sops')
```
Comment (mloubout) [inline devito/operator/operator.py:922]: `cire-maxalias` as well

Comment (FabioLuporini) [reply devito/operator/operator.py:922]: added commit already in follow-up PR
